pkg/counter/cache.go

```diff
@@ -0,0 +1,162 @@
+func IncrementMasterUserRecordCount() {
+	write(func() {
+		cachedCounts.MasterUserRecordCount++
+	})
+}
+
+// DecrementMasterUserRecordCount decreases the number of MasterUserRecord in the cached counter
+func DecrementMasterUserRecordCount(log logr.Logger) {
+	write(func() {
+		if cachedCounts.MasterUserRecordCount != 0 {
+			cachedCounts.MasterUserRecordCount--
+		} else {
+			log.Error(fmt.Errorf("the count of MasterUserRecords is zero"),
+				"unable to decrement the number of MasterUserRecords")
+		}
+	})
+}
+
+// IncrementUserAccountCount increments the number of UserAccount fo the given member cluster in the cached counter
+func IncrementUserAccountCount(clusterName string) {
+	write(func() {
+		cachedCounts.UserAccountsPerClusterCounts[clusterName]++
+	})
+}
+
+// DecrementUserAccountCount decreases the number of UserAccount fo the given member cluster in the cached counter
+func DecrementUserAccountCount(log logr.Logger, clusterName string) {
+	write(func() {
+		if cachedCounts.UserAccountsPerClusterCounts[clusterName] != 0 {
+			cachedCounts.UserAccountsPerClusterCounts[clusterName]--
+		} else {
+			log.Error(fmt.Errorf("the count of UserAccounts is zero"),
+				"unable to decrement the number of UserAccounts for the given cluster", "cluster", clusterName)
+		}
+	})
+}
+
+// GetCounts returns Counts struct containing number of MURs and number of UserAccounts per member cluster.
+// If the counter is not yet initialized, then it returns error
+func GetCounts() (Counts, error) {
+	cachedCounts.RLock()
+	defer cachedCounts.RUnlock()
+	if !cachedCounts.initialized {
+		return cachedCounts.Counts, fmt.Errorf("counter is not initialized")
+	}
+	return cachedCounts.Counts, nil
+}
+
+// Synchronize synchronizes the content of the ToolchainStatus with the cached counter
+//
+// If the counter hasn't been initialized yet, then it adds the actual cached count
+// to the one taken from ToolchainStatus and marks the cached as initialized
+//
+// If the ToolchainStatus doesn't contain any MUR counts, then it lists all existing
+// MURs, counts them and stores in both cache and ToolchainStatus object
+//
+// If the cached counter is initialized and ToolchainStatus contains already some numbers
+// then it updates the ToolchainStatus numbers with the one taken from the cached counter
+func Synchronize(cl client.Client, toolchainStatus *v1alpha1.ToolchainStatus) error {
+	cachedCounts.Lock()
+	defer cachedCounts.Unlock()
+	if toolchainStatus.Status.HostOperator == nil || toolchainStatus.Status.HostOperator.CapacityUsage.MasterUserRecordCount == 0 {
+		if err := loadCurrentNumbers(cl, toolchainStatus.Namespace); err != nil {
+			return err
+		}
+	}
+	if toolchainStatus.Status.HostOperator == nil {
+		toolchainStatus.Status.HostOperator = &v1alpha1.HostOperatorStatus{}
+	}
+	if !cachedCounts.initialized {
+		cachedCounts.MasterUserRecordCount += toolchainStatus.Status.HostOperator.CapacityUsage.MasterUserRecordCount
```

Comment:
why `cachedCounts.MasterUserRecordCount += ` and not just `cachedCounts.MasterUserRecordCount = ` ?

Reply:
because I want to cover the case when it could happen that there is a new MUR or UserAccount created in the small tiny gap before the cache is initialized. Then the cache would already contain some number with need to add to the one in `ToolchainStatus`.